lib/mapObjects/CommonConstructors.cpp

```diff
@@ -291,6 +291,69 @@ void BoatInstanceConstructor::configureObject(CGObjectInstance * object, CRandom
 
 }
 
+void MarketInstanceConstructor::initTypeData(const JsonNode & input)
+{
+	for(auto & element : input["modes"].Vector())
+	{
+		if(MappedKeys::MARKET_NAMES_TO_TYPES.count(element.String()))
+			marketModes.insert(MappedKeys::MARKET_NAMES_TO_TYPES.at(element.String()));
+	}
+	
+	marketEfficacy = input["efficacy"].isNull() ? -1 : input["efficacy"].Integer();
+	predefinedOffer = input["offer"];
+	
+	title = input["title"].String();
+	speech = input["speech"].String();
+}
+
+CGObjectInstance * MarketInstanceConstructor::create(std::shared_ptr<const ObjectTemplate> tmpl) const
+{
+	CGMarket * market = nullptr;
+	if(marketModes.size() == 1)
+	{
+		switch(*marketModes.begin())
+		{
+			case EMarketMode::ARTIFACT_RESOURCE:
+			case EMarketMode::RESOURCE_ARTIFACT:
+				market = new CGBlackMarket;
+				break;
+				
+			case EMarketMode::RESOURCE_SKILL:
+				market = new CGUniversity;
+				break;
+		}
+	}
+	
+	if(!market)
+		market = new CGMarket;
+	
+	preInitObject(market);
+
+	if(tmpl)
+		market->appearance = tmpl;
+	market->marketModes = marketModes;
+	if(marketEfficacy >= 0)
+		market->marketEfficacy = marketEfficacy;
```

Comment:
Perhaps this should be obligatory member and set unconditionally?

Reply:
no, it will be -1 if field is missing. Then 5 default value will be used

Reply:
Any need for this logic?

Reply:
I think this parameter is not supposed to be used (almost). For all known objects value is 5. Then better to allow omitting parameter making it optional

Reply:
But this is not how all our modding system is written.
We *always* specify all parameters. We have almost no conditional branches in objects loading code.
Even if all H3 objects use the same value we have it specified in all our objects

Reply:
it's not true. Most of parameters in every loader are optional. Open any mod - obviously, modders specify only parameters they need (and it would be completely unfriendly to make the specify everything).

So it's obviously and user-oriented that some parameters must be optional. So we provide flexibility for modders to specify some rare property but please let's not make them specify parameter which they don't need. Such approaches remind me about freaky downloading web-portals, where you must watch 60 sec at advertisement before link appears))

Reply:
However we always do that via zero-init - all such parameters are specifically designed to have false/0/empty string as "reasonable default".

Parameters where such zero-init is not a "reasonable default" always have explicit value in config.

Optional values often lead to an opposite effect - where modder forgets to put parameter that is needed for his object or where modder simply not aware of existence of such parameter (since making parameter as required would report it in validation)

Reply:
Exactly, that's why reasonable default is -1 here, not 5

> Optional values often lead to an opposite effect

and that's totally fine if people aren't aware about something. Definitely better that awareness by force.
Can be solved in "soft" way - documentation, examples, and all-defined approach for embedded objects (described in config folder which can be used as starting point for mods)

Reply:
If you insist on keeping this then at the very least please define all of this in one place, with one conditional check, e.g. in MarketConstuctor initializer.
Currently you have:
- marketConstructor initializing this property to -1 if not found in json
- marketConstructor assigning this property if above 0
- market itself having default value set to 5

Why not simply write something like
```c++
if (!json["efficiency"].isNull())
    efficiency = json["efficiency"].Integer()
else
    efficiency = 5;
```
Instead of spreading this through the code? Right now you have 3 different constants in 3 different locations when it coult be 1 constant in 1 location.

> Can be solved in "soft" way - documentation, examples, and all-defined approach for embedded objects (described in config folder which can be used as starting point for mods)

Modders must be aware of these documentation & examples. They are often not.
This documentation and examples must be maintained. They are often not.

I really like our validation system since it basically enforces everything to be up to date - it checks for any error made by us (such as not tracked field), for any error made by modder (such as missing field) and as result - always up to date. This is way more reliable than documentation & examples.

If I eventually get to my "modding tool" idea it will be based on our validator & will act as source of documentation & examples specifically because it is up to date by design.

Reply:
understood, you are talking not about optional/mandatory parameter, but about initialzation out of loader

Reply:
Kind of both.
Current approach is unnecessarily complex, with developer having to be aware of 3 places responsible for loading.
Reducing it to one conditional in one location is an improvement, but still (in my opinion) unnecessary complexity.

It is fine now since this is basically the only parameter that markets can have. But when you have 10-20 parameters having if-else block for each one makes code too hard to understand.

Reply:
I suppose can be solved by overloading cast methods in JsonNode:
```c++
int JsonNode::Integer() const; //mandatory parameter, exception if missing
int JsonNode::Integer(int defaulValue) const; //optional parameter, exception-safe
```

Reply:
Also an option, although right now we also have an advantage that default values are clear to modders - since missing values are always treated as false/0/empty string. 

Such change also means that we now need to show default value in docs & make sure that those default parameters are up to date with actual code.